### PR TITLE
Merge from google (via MOE)

### PR DESCRIPTION
**Pull request by cgruber**
Improvements to validations, `@Subcomponent`s, easier map binding key use, and other various fixes.

### CODE REVIEW

File: compiler/src/test/java/dagger/internal/codegen/SubcomponentValidationTest.java [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+        "@Subcomponent(modules = ModuleWithParameters.class)",
+        "interface ChildComponent {}");
+    JavaFileObject moduleFile = JavaFileObjects.forSourceLines("test.ModuleWithParameters",
+        "package test;",
+        "",
+        "import dagger.Module;",
+        "",
+        "@Module",
+        "final class ModuleWithParameters {",
+        "  ModuleWithParameters(Object whatever) {}",
```
Comment (adriancole) [inline compiler/src/test/java/dagger/internal/codegen/SubcomponentValidationTest.java]: prefer `whatev` as it is hipper

Comment (cgruber) [inline compiler/src/test/java/dagger/internal/codegen/SubcomponentValidationTest.java]: Noted.

**Comment by cgruber** [comment]
The fixes for this build are all working their way through internal review.  If we can get them all done and synced out, we should still be able to release today.

**Comment by gk5885** [comment]
LGTM

File: compiler/src/main/java/dagger/internal/codegen/ErrorMessages.java [hunk 2]
```diff
@@ -195,10 +195,10 @@
       "This type supports members injection but cannot be implicitly provided.";
 
   static final String MEMBERS_INJECTION_WITH_RAW_TYPE =
-      "%s has type parameters, cannot members inject the raw type.";
+      "%s has type parameters, cannot members inject the raw type. via:\n%s";
 
   static final String MEMBERS_INJECTION_WITH_UNBOUNDED_TYPE =
-      "Type parameters must be bounded for members injection.";
+      "Type parameters must be bounded for members injection. (%s required by %s, via:\n%s";
```
Comment (JakeWharton) [inline compiler/src/main/java/dagger/internal/codegen/ErrorMessages.java]: Missing closing paren

Comment (cgruber) [inline compiler/src/main/java/dagger/internal/codegen/ErrorMessages.java]: Addressed in a forthcoming commit.

**Pull request merged**